### PR TITLE
The ANTLR-isation of the lexical grammar. The C# "lexical" grammar bo…

### PR DESCRIPTION
**Pull request by Nigel-Ecma**
…th tokenises and parses the input and this

requires using both lexical and parser ANTLR rules while producing a lexer which produces a token stream and
not some AST of the input!

In this stage of ANTLR-isation the pre-processor is not tackled; the pre-processor definition is a context-free
grammar which is atypical for a lexical grammar and defining it with lexical rules does not produce the expected
result... That said the semantics of the pre-processor can be acheived in an ANTLR lexical grammar which includes
predicates/actions and that will be explored in a subsequent stage. The grammar in this stage can be used along
with a factored syntactical grammar to lex & parse C# source which does not include pre-processor constructs.

The majority of changes are to lexical-structure.md, many of the other files are only altered as the lexical
rule *Identifier* has been replaced by the parser rule *identifier*.

I will add some comments within the PR to explain why it is the way it is...

### CODE REVIEW

File: standard/grammar.md [hunk 1]
```diff
@@ -19,23 +19,23 @@ ASTERISK : '*' ;
 SLASH    : '/' ;
 
 // Source: §7.3.1 General
-Input
-    : Input_Section?
+input
```
Comment (Nigel-Ecma) [inline standard/grammar.md:22]: This file is largely auto-generated on GitHub, the changes here (a) reflect changes elsewhere and (b) are incomplete. Once the GitHub tooling runs they will be replaced. In other words don't review this file!

File: standard/lexical-structure.md [hunk 2]
```diff
@@ -26,23 +26,29 @@ All terminal characters are to be understood as the appropriate Unicode characte
 
 ### 7.2.2 Grammar notation
 
-The lexical and syntactic grammars are presented in Backus-Naur form using the notation of the ANTLR grammar tool.
+The lexical and syntactic grammars are presented in the ANTLR grammar tool’s Extended Backus-Naur form.
+
+While the ANTLR notation is used this Standard does not present a complete ANTLR-ready "reference grammar" for C#; writing a lexer and parser, either by hand or using a tool such as ANTLR, is outside the scope of a language specification. With that qualification, this Standard attempts to minmise the gap between the specified grammar and that required to build a lexer and parser in ANTLR with the notable exception of the preprocessor ([§7.5](lexical-structure.md#75-pre-processing-directives) which requires more substantial work to fit into the ANTLR model.
```
Comment (Nigel-Ecma) [inline standard/lexical-structure.md]: Normative, informative, or a mix?

File: standard/lexical-structure.md [hunk 3]
```diff
@@ -26,23 +26,29 @@ All terminal characters are to be understood as the appropriate Unicode characte
...
+
+ANTLR distinguishes between lexical and syntactic, termed parser by ANTLR, grammars in its notation by starting lexical rules with an initial uppercase letter and parser rules with an initial lowercase letter.
+
+*Note*: The C# lexical grammar ([§7.2.3](lexical-structure.md#723-lexical-grammar)) and syntactic grammar ([§7.2.4](lexical-structure.md#724-syntactic-grammar)) are not in exact correspondence with the ANTLR division into lexical and parser grammers. This small mismatch means that some ANTLR parser rules are used when specifying the C# lexical grammar. *end note*
```
Comment (Nigel-Ecma) [inline standard/lexical-structure.md]: Or put another way the C# "lexical" grammar is a context-free grammar not a regular one... Nothing wrong with that per se but it's not the model a tool like ANTLR follows.

I've added a number of notes (which are informative rather than normative) thoughout, the number & content of these should be reviewed.

File: standard/lexical-structure.md [hunk 4]
```diff
@@ -26,23 +26,29 @@ All terminal characters are to be understood as the appropriate Unicode characte
...
 
 ### 7.2.3 Lexical grammar
 
 The lexical grammar of C# is presented in [§7.3](lexical-structure.md#73-lexical-analysis), [§7.4](lexical-structure.md#74-tokens), and [§7.5](lexical-structure.md#75-pre-processing-directives). The terminal symbols of the lexical grammar are the characters of the Unicode character set, and the lexical grammar specifies how characters are combined to form tokens ([§7.4](lexical-structure.md#74-tokens)), white space ([§7.3.4](lexical-structure.md#734-white-space)), comments ([§7.3.3](lexical-structure.md#733-comments)), and pre-processing directives ([§7.5](lexical-structure.md#75-pre-processing-directives)).
 
-Every compilation unit in a C# program shall conform to the *Input* production of the lexical grammar ([§7.3.1](lexical-structure.md#731-general)).
+Many of the terminal symbols of the syntactic grammar are not defined explicitly as tokens in the lexical grammar. Rather advantage is taken of the ANTLR behaviour that literal strings in the grammar are extracted as implicit lexical tokens; this allows keywords, operators, etc. to be represented in the grammar by their literal representation rather than a token name.
```
Comment (Nigel-Ecma) [inline standard/lexical-structure.md]: That literal strings in the grammar become implicit lexical tokens was also the case with the original grammar. The impact of this was never spelled out in any way and indeed the lexical rules around identifiers were described as though the implicitly defined keyword tokens didn't exist. As we are now using the language of a real parser generator this detail has been added.

File: standard/lexical-structure.md [hunk 5]
```diff
@@ -99,31 +105,33 @@ ASTERISK : '*' ;
...
-Input_Element
+input_element
     : Whitespace
     | Comment
-    | Token
+    | token
     ;
 ```
 
+> *Note*: The above grammar is described by ANTLR parsing rules, it defines the lexical structure of a C# compilation unit and not lexical tokens. *end note*
```
Comment (Nigel-Ecma) [inline standard/lexical-structure.md]: Really this grammar is not about tokens (lexical) but about arrangements of tokens (parser). A lexer might implement them to check the structure of the input but there is no need and in ANTLR they are simply ignored as there is no path from the start symbol of the parser grammar to them. However they are the only place that *PP_Directive* is referenced, but that is no ANTLR-ised yet...

File: standard/lexical-structure.md [hunk 6]
```diff
@@ -132,6 +140,12 @@ When several lexical grammar productions match a sequence of characters in a com
 
 > *Example*: The character sequence `//` is processed as the beginning of a single-line comment because that lexical element is longer than a single `/` token. *end example*
 
+The rules of the lexical grammar are ordered top-to-bottom, with the first matching rule determining the recognised token. Implicit rules ([§7.2.3](lexical-structure.md#723-lexical-grammar)) derived from literal strings in the grammar are ordered before the explicit rules. This ordering simplifies the grammar.
```
Comment (Nigel-Ecma) [inline standard/lexical-structure.md:151]: This top-to-bottom ordering is not mentioned in the original grammar but is how lexical grammars usually work – lexical rules are ordered while parser rules form a tree rooted in a start symbol. It is important to mention it once the implicit rules for keywords etc. are introduced.

File: standard/lexical-structure.md [hunk 7]
```diff
@@ -324,64 +340,94 @@ Multiple translations are not performed. For instance, the string literal `"\u00
 The rules for identifiers given in this subclause correspond exactly to those recommended by the Unicode Standard Annex 15 except that underscore is allowed as an initial character (as is traditional in the C programming language), Unicode escape sequences are permitted in identifiers, and the "`@`" character is allowed as a prefix to enable keywords to be used as identifiers.
 
 ```ANTLR
-Identifier
+identifier
```
Comment (Nigel-Ecma) [inline standard/lexical-structure.md:385]: This part of the grammar has been re-formulated noting (a) the impact of the implicit rules generated from literal strings in the grammar and (b) to leverage those so "contextual keywords" are recognised *by the grammar* as identifiers when not being contextually – there might be cases when a contextual keyword is parsed as an identifier even though its in a contextual position but that can easily be handled by the compiler.

File: standard/lexical-structure.md [hunk 8]
```diff
@@ -324,64 +340,94 @@ Multiple translations are not performed. For instance, the string literal `"\u00
...
+    : Simple_Identifier
+    | contextual_keyword
+    ;
+
+Simple_Identifier
```
Comment (Nigel-Ecma) [inline standard/lexical-structure.md:390]: You might wonder why *Simple_Identifier* exists and its simply not folded into *identifier*. So that the main lexical rules can be represented as a lexical grammar, rather than a combined grammar, in ANTLR the C# "lexical" rules are split into two ANTLR grammars (a) a lexical one for lexical rules and (b) a parser one for the parser rules. This arrangement will allow the pre-processor to be implemented as such an implementation is likely to rely on features only ANTLR lexical-only grammars can use. The rule `Simple_Identifer` provides the bridge between the lexer and parser rules of the "lexical" grammar... Now you know why I didn't explain that in the text, too tortuous! And its not as though the grammar doesn't have plenty of redundant rules for the purposes of exposition already, and the next one is immediately below...

File: standard/lexical-structure.md [hunk 9]
```diff
@@ -324,64 +340,94 @@ Multiple translations are not performed. For instance, the string literal `"\u00
...
     | Combining_Character
     | Formatting_Character
     ;
 
-Letter_Character
-    : [\p{L}\p{Nl}]     // category Letter, all subcategories; category Number, subcategory letter
-    | Unicode_Escape_Sequence { IsLetterCharacter() }?
+fragment Letter_Character
+    : [\p{L}\p{Nl}]           // category Letter, all subcategories; category Number, subcategory letter
+    | Unicode_Escape_Sequence // only escapes for categories L & Nl allowed, see note below
```
Comment (Nigel-Ecma) [inline standard/lexical-structure.md:429]: Should we include the semantic predicates? I'm in two (at least) minds, so we get both a comment (as in the original), a predicate, and a note below explaining they are not normative...

File: standard/lexical-structure.md [hunk 10]
```diff
@@ -324,64 +340,94 @@ Multiple translations are not performed. For instance, the string literal `"\u00
...
 
-> *Note*: For information on the Unicode character classes mentioned above, see *The Unicode Standard*. *end note*
+> *Note*:
+
+> 1. For information on the Unicode character classes mentioned above, see *The Unicode Standard*. *end note*
+
+> 2. The ANTLR semantic predicates above: `IsLetterCharacter`, `IsCombiningCharacter`, `IsDecimalDigitCharacter`, `IsConnectingCharacter` and `IsFormattingCharacter`; are informative *only*. How a compiler enforces the restriction on the allowable
+*Unicode_Escape_Sequence* values is an implementation issue. *end note*
+
+> 3. The explicit rules for identifiers depend on the implicit rules introduced by literal strings in the grammar:
```
Comment (Nigel-Ecma) [inline standard/lexical-structure.md:465]: This attempts to clarify how the implicit and explicit rules interact to tokenise identifiers and [contextual] keywords, does it? Please review carefully.

File: standard/lexical-structure.md [hunk 11]
```diff
@@ -758,26 +813,28 @@ There are several kinds of operators and punctuators. Operators are used in expr
...
     ;
 
-Right_Shift_Assignment
+right_shift_assignment
     : '>' '>='
     ;
 ```
 
-*Right_Shift* is made up of the two tokens `>` and `>`. Similarly, *Right_Shift_Assignment* is made up of the two tokens `>` and `>=`. Unlike other productions in the syntactic grammar, no characters of any kind (not even whitespace) are allowed between the two tokens in each of these productions. These productions are treated specially in order to enable the correct handling of *type_parameter_lists* ([§15.2.3](classes.md#1523-type-parameters)). 
+> *Note*: `right_shift` and `right_shift_assignment` are parser rules as they do not introduce a new token kind but represent a sequence of two tokens. The `operator_or_punctuator` rule exists for descriptive purposes only and is not used elsewhere in the grammar. *end note*
```
Comment (Nigel-Ecma) [inline standard/lexical-structure.md:875]: Not said is that in the original grammar these two rules didn't really work as they would both have produced a single token...

File: standard/lexical-structure.md [hunk 12]
```diff
@@ -787,14 +844,16 @@ The pre-processing directives provide the ability to skip conditionally sections
 
 > *Note*: The term "pre-processing directives" is used only for consistency with the C and C++ programming languages. In C#, there is no separate pre-processing step; pre-processing directives are processed as part of the lexical analysis phase. *end note*
 
+> *Note*: The pre-processor grammar, though defined using ANTLR lexical rules, does not define any lexical tokens in the language and is not “ANTLR ready”. Pre-processing involves processing the input and modifying the token stream produced by the lexer, this is not represented in this grammar and is part of compiler implementation and thus out of scope for this language specification. *end note*
```
Comment (Nigel-Ecma) [inline standard/lexical-structure.md]: This is the big bit that hasn't been ANTLR-ised. I did look as to whether these rules could be treated as lexical ones which produce a single "PP Directive" token, but that is thwarted by the context-free specification...

Though I think I have a proposal as to how to handle this as any PR will build on this one and I've no idea how to do that in GitHub I've chosen to leave this until we've decided whether to accept this PR and its merged into the main draft-v6 branch. If you would like to know that proposal to help decide on *this* PR request it on the ECMA list and I'll explain once I've tested it.

**Comment by RexJaeschke** [comment]
@Nigel-Ecma I looked over this, but given my lack of background w.r.t formal grammars, I'm not following a lot of the explanation. As such, I removed myself from the reviewer's list.

BTW, I think that in at least one place you have a Note that *doesn't* start on a new line with a preceding `>`. We previously agreed that Notes can no longer be at the end of ordinary paragraphs.

**Comment by Nigel-Ecma** [comment]
@RexJaeschke – I think I’ve fixed all the inline Notes that I added (there was more than one); in looking for them I did a global search and turned up two elsewhere. I have not fixed them in the draft PR as they are not connected:

basic-concepts.md, line 712
conversions.md, line 444

File: standard/lexical-structure.md [hunk 13]
```diff
@@ -26,23 +26,29 @@ All terminal characters are to be understood as the appropriate Unicode characte
 
 ### 7.2.2 Grammar notation
 
-The lexical and syntactic grammars are presented in Backus-Naur form using the notation of the ANTLR grammar tool.
+The lexical and syntactic grammars are presented in the ANTLR grammar tool’s Extended Backus-Naur form.
```
Comment (jskeet) [inline standard/lexical-structure.md]: Are we using curly quotes in the standard in general? I'd normally expect an ASCII `'`.

File: standard/lexical-structure.md (hunk 2 above)
Comment (jskeet) [reply standard/lexical-structure.md]: Nit: minmise => minimize

I think it's reasonable to keep this part normative. Will review the mix as I go along.

File: standard/lexical-structure.md (hunk 4 above)
Comment (jskeet) [reply standard/lexical-structure.md]: Nit: behavior => behavior (it pains me, but...) - same later on.

File: standard/lexical-structure.md (hunk 6 above)
Comment (jskeet) [reply standard/lexical-structure.md:151]: +1 for mentioning it explicitly.

File: standard/lexical-structure.md (hunk 5 above)
Comment (jskeet) [reply standard/lexical-structure.md]: Both the standard change and the comment have gone so far over my head, they've left earth's atmosphere. I think that's fine - but if you expect a useful review of this little bit, please let me know :)

File: standard/lexical-structure.md [hunk 14]
```diff
@@ -324,64 +340,94 @@ Multiple translations are not performed. For instance, the string literal `"\u00
...
+         { IsFormattingCharacter() }?
     ;
 ```
 
-> *Note*: For information on the Unicode character classes mentioned above, see *The Unicode Standard*. *end note*
+> *Note*:
+
+> 1. For information on the Unicode character classes mentioned above, see *The Unicode Standard*. *end note*
+
+> 2. The ANTLR semantic predicates above: `IsLetterCharacter`, `IsCombiningCharacter`, `IsDecimalDigitCharacter`, `IsConnectingCharacter` and `IsFormattingCharacter`; are informative *only*. How a compiler enforces the restriction on the allowable
```
Comment (jskeet) [inline standard/lexical-structure.md]: I'd put that list of predicates into parentheses instead:

```text
The ANTLR semantic predicates above (`IsLetterCharacter`, `IsCombiningCharacter`, `IsDecimalDigitCharacter`, `IsConnectingCharacter` and `IsFormattingCharacter`) are informative *only*.
```

File: standard/lexical-structure.md [hunk 15]
```diff
@@ -324,64 +340,94 @@ Multiple translations are not performed. For instance, the string literal `"\u00
...
+*Unicode_Escape_Sequence* values is an implementation issue. *end note*
+
+> 3. The explicit rules for identifiers depend on the implicit rules introduced by literal strings in the grammar:
+>> - Keywords and contextual keywords occur in the grammar as literal strings and from these implicit lexical token rules are created which are ordered before the explicit lexical rules ([§7.2.3](lexical-structure.md#723-lexical-grammar)).
```
Comment (jskeet) [inline standard/lexical-structure.md]: What's the purpose of the `>>` here? That's Markdown I haven't seen before.

Comment (jskeet) [reply standard/lexical-structure.md]: In terms of the text itself, I think I'd find this easier to read as three sentences:

> Keywords and contextual keywords occur in the grammar as literal strings. Implicit lexical token rules are created from these literal strings. The implicit rules are ordered before the explicit lexical rules ([§7.2.3](lexical-structure.md#723-lexical-grammar)).

File: standard/lexical-structure.md [hunk 16]
```diff
@@ -324,64 +340,94 @@ Multiple translations are not performed. For instance, the string literal `"\u00
...
+>> - Therefore fragment `Available_Identifier` will not match keywords or contextual keywords as the lexical rules for those precede it.
```
Comment (jskeet) [inline standard/lexical-structure.md]: Having "Therefore" at the start of the bullet feels a little odd to me. How about

"The ordering of the token rules means that fragment [...]"

?

File: standard/lexical-structure.md [hunk 17]
```diff
@@ -324,64 +340,94 @@ Multiple translations are not performed. For instance, the string literal `"\u00
...
+>> - Fragment `Escaped_Identifier` will effectively include keywords and contextual keywords as part of the longer token starting with an `@`.
+>> - `identifier` is a parser rule, as are those for `keyword` and `contextual_keyword` ([§7.4.4](lexical-structure.md#744-keywords)), as they do not define a new token kinds but rather represent a grouping of other tokens.
```
Comment (jskeet) [inline standard/lexical-structure.md]: Would "as are `keyword` and `contextual_keyword`" work just as well? It's not clear to me what the "those for" means or why it's different for `keyword`/`contextual_keyword` vs `identifier`.

File: standard/lexical-structure.md [hunk 18]
```diff
@@ -459,6 +505,8 @@ Contextual_Keyword
     ;
 ```
 
+> *Note*: The rules `keyword` and `contextual_keyword` are parser rules as they do not introduce new token kinds. Every keyword and contextual keyword is defined by an implicit lexical rule as they occur as literal strings in the grammar ([§7.2.3](lexical-structure.md#723-lexical-grammar)). *end note*
```
Comment (jskeet) [inline standard/lexical-structure.md]: Suggested pluralization consistency:

"All keywords and contextual keywords are defined by implicit lexical rules as they occur as literal strings in the grammar"

File: standard/lexical-structure.md [hunk 19]
```diff
@@ -758,26 +813,28 @@ There are several kinds of operators and punctuators. Operators are used in expr
...
 
-> *Note*: Prior to the addition of generics to C#, `>>` and `>>=` were both single tokens. However, the syntax for generics uses the `<` and `>` characters to delimit type parameters and type arguments. It is often desirable to use nested constructed types, such as `List<Dictionary<string`, `int>>`. Rather than requiring the programmer to separate the `>` and `>` by a space, the definition of the two *Operator_Or_Punctuator*s was changed.
+> *Note*: Prior to the addition of generics to C#, `>>` and `>>=` were both single tokens. However, the syntax for generics uses the `<` and `>` characters to delimit type parameters and type arguments. It is often desirable to use nested constructed types, such as `List<Dictionary<string`, `int>>`. Rather than requiring the programmer to separate the `>` and `>` by a space, the definition of the two *operator_or_punctuator*s was changed.
```
Comment (jskeet) [inline standard/lexical-structure.md]: I don't think we want the backticks within the constructed type here - this is a bug in the existing Markdown which we didn't catch in previous review.

File: standard/lexical-structure.md (hunk 12 above)
Comment (jskeet) [reply standard/lexical-structure.md]: Suggest replacing the curly-double-quotes with straight ones. But other than that, I'm fine with the paragraph and your comment :)

File: standard/lexical-structure.md [hunk 20]
```diff
@@ -1200,16 +1259,16 @@ The `#pragma` preprocessing directive is used to specify contextual information
...
 
-Pp_Pragma_Text
+PP_Pragma_Text
     : New_Line
     | Whitespace Input_Character* New_Line
     ;
 ```
 
-The *Input_Character*s in the *Pp_Pragma-Text* are interpreted by the compiler in an implementation-defined manner. The information supplied in a `#pragma` directive shall not change program semantics. A `#pragma` directive shall only change compiler behavior that is outside the scope of this language specification. If the compiler cannot interpret the *Input_Character*s, the compiler can produce a warning; however, it shall not produce a compile-time error.
+The *Input_Character*s in the *PP_Pragma-Text* are interpreted by the compiler in an implementation-defined manner. The information supplied in a `#pragma` directive shall not change program semantics. A `#pragma` directive shall only change compiler behavior that is outside the scope of this language specification. If the compiler cannot interpret the *Input_Character*s, the compiler can produce a warning; however, it shall not produce a compile-time error.
```
Comment (jskeet) [inline standard/lexical-structure.md:1315]: Should the - in *PP_Pragma-Text* be an underscore?

File: standard/lexical-structure.md [hunk 21]
```diff
@@ -144,7 +146,9 @@ The rules of the lexical grammar are ordered top-to-bottom, with the first match
 
 > *Example*: As keyword tokens are defined implicitly by literal strings in the grammar, and implicit rules are ordered before explicit ones, the rule `Available_Identifier` ([§7.4.3](lexical-structure.md#743-identifiers)) will never match a keyword. *end example*
 
-Some tokens are defined by a set of lexical rules; a main rule and one or more sub-rules. The latter are marked in the grammer by `fragment` to indicate the rule does defines part of another token. Fragment rules are not considered in the top-to-bottom ordering of lexical rules. (*Note*: In ANTLR `fragment` is a keyword which produces the same behaviour defined here. *end note*)
+Some tokens are defined by a set of lexical rules; a main rule and one or more sub-rules. The latter are marked in the grammer by `fragment` to indicate the rule does defines part of another token. Fragment rules are not considered in the top-to-bottom ordering of lexical rules.
```
Comment (jskeet) [inline standard/lexical-structure.md]: Nit: grammer => grammar

File: standard/lexical-structure.md [hunk 22]
```diff
@@ -285,18 +291,50 @@ Whitespace
...
+    | 'readonly'   | 'ref'       | 'remove'     | 'return'    | 'sbyte'
+    | 'sealed'     | 'select'    | 'set'        | 'short'     | 'sizeof'
+    | 'stackalloc' | 'static'    | 'string'     | 'struct'    | 'switch'
+    | 'this'       | 'throw'     | 'true'       | 'try'       | 'typeof'
+    | 'uint'       | 'ulong'     | 'unchecked'  | 'unsafe'    | 'ushort'
+    | 'using'      | 'value'     | 'var'        | 'virtual'   | 'void'
+    | 'volatile'   | 'where'     | 'while'      | 'yield'
+    ;
+
+fragment Any_Operator_Or_Punctuator
```
Comment (jskeet) [inline standard/lexical-structure.md:327]: I seem to remember a reference to Operator_Or_Punctuator in a note, but I can't see a change to that note here.

File: standard/lexical-structure.md [hunk 23]
```diff
@@ -909,10 +947,15 @@ Pre-processing directives are not tokens and are not part of the syntactic gramm
...
 ```ANTLR
-Conditional_Symbol
-    : Identifier_Or_Keyword { IsNotTrueOrFalse() }?
+fragment Conditional_Symbol
+    : Basic_Identifier           // excluding TRUE and FALSE
+         { IsNotTrueOrFalse() }? // see note below
     ;
 ```
+
+> *Note* The ANTLR semantic predicate above: `IsNotTrueOrFalse`; is informative *only*. How a compiler enforces the restriction on the allowable
```
Comment (jskeet) [inline standard/lexical-structure.md]: As before, I'd reword this a bit:

"The ANTLR semantic predicate above (`IsNotTrueOrFalse`) is informative *only*."

**Review by jskeet — COMMENTED** [review]
I'm not really qualified to comment on the validity of the overall proposal, but I have complete confidence in Nigel's efforts given his description in the previous meeting.

Not being qualified hasn't stopped me from adding various nit-picks and suggestions though :)

File: standard/lexical-structure.md (hunk 15 above)
Comment (Nigel-Ecma) [reply standard/lexical-structure.md]: >> is double-indent, saves you having to insert spaces to produce nested lists. I've switched to a single > and more spaces assuming that's the approach elsewhere. Switch to 3 sentences but started the last with "These" as it read better to me.

Comment (Nigel-Ecma) [reply standard/lexical-structure.md]: And the double indent/quote worked in the comment ;-)

File: standard/lexical-structure.md (hunk 16 above)
Comment (Nigel-Ecma) [reply standard/lexical-structure.md]: We start bullets with "Otherwise", I think I'm going to disagree on this one, let's see where the consensus lands.

File: standard/lexical-structure.md (hunk 20 above)
Comment (Nigel-Ecma) [reply standard/lexical-structure.md:1315]: I spotted this pre-existing typo when doing #342, I'll fix it here as well. There might be other fixes I found and made in #342, someone with good Git-fu I'm sure could merge them if we go with #339. I'm making all the fixes you found in both branches.

File: standard/lexical-structure.md (hunk 2 above)
Comment (Nigel-Ecma) [reply standard/lexical-structure.md]: Catching my typos is not a nit ;-) Fixed.

File: standard/lexical-structure.md (hunk 4 above)
Comment (Nigel-Ecma) [reply standard/lexical-structure.md]: Yup, I'm used to correct (British/New Zealand) spelling ;-)

File: standard/lexical-structure.md (hunk 13 above)
Comment (Nigel-Ecma) [reply standard/lexical-structure.md]: All the quotes (some historical) in *this* file have now been straightened, I leave the rest of the Standard to others. I trust that the Word convertor will put them back in.

File: standard/lexical-structure.md (hunk 14 above)
Comment (Nigel-Ecma) [reply standard/lexical-structure.md]: Done

File: standard/lexical-structure.md (hunk 17 above)
Comment (Nigel-Ecma) [reply standard/lexical-structure.md]: Dropped

File: standard/lexical-structure.md (hunk 18 above)
Comment (Nigel-Ecma) [reply standard/lexical-structure.md]: Done

File: standard/lexical-structure.md (hunk 19 above)
Comment (Nigel-Ecma) [reply standard/lexical-structure.md]: Well spotted, yes an old typo, fixed.

File: standard/lexical-structure.md (hunk 21 above)
Comment (Nigel-Ecma) [reply standard/lexical-structure.md]: Corrected

File: standard/lexical-structure.md (hunk 23 above)
Comment (Nigel-Ecma) [reply standard/lexical-structure.md]: Changed

**Comment by Nigel-Ecma** [comment]
Supplanted by #342
